Implement accepting dual-funded channels without contributing

dunxen opened the pull request:
We split this out from #2302 for easier review and to address the common non-public API parts of the V2 channel establishment implementation.

This will allow the holder to be an acceptor, but not initiator of V2 channels. We also don't expose an API for contributing to an inbound channel.

The functionality to initiate V2 channels and fund inbound channels forms part of #2302.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+		// if let Ok((tx_constructor, msg)) = 
+		match InteractiveTxConstructor::new(
+			entropy_source, self.context().channel_id(),
+			self.dual_funding_context_mut().funding_feerate_sat_per_1000_weight,
```
```suggestion
			self.dual_funding_context().funding_feerate_sat_per_1000_weight,
```

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+			entropy_source, self.context().channel_id(),
+			self.dual_funding_context_mut().funding_feerate_sat_per_1000_weight,
+			holder_node_id, self.context().counterparty_node_id, self.is_initiator(),
+			self.dual_funding_context_mut().funding_tx_locktime,
+			self.dual_funding_context_mut().our_funding_inputs.clone().unwrap_or(vec![]), funding_outputs,
```
```suggestion
			self.dual_funding_context().funding_tx_locktime,
			self.dual_funding_context().our_funding_inputs.clone().unwrap_or(vec![]), funding_outputs,
```

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+			);
+		}
+
+		// if let Ok((tx_constructor, msg)) = 
```
Prob. outdated, remove

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+			Err(err) => {
+				Err(APIError::APIMisuseError { err:
+					match err {
+				        AbortReason::MissingFundingOutput => "Funding output missing".into(),
```
Nit: Please check your white space settings, there are mixed tabs and spaces used in these lines

dunxen replied:
The settings are fine for my editor, it's when rust-analyzer fills in branches that it defaults to spaces. I'll have a look at changing its config.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+	fn tx_add_input(&mut self, msg: &msgs::TxAddInput) -> InteractiveTxMessageSendResult {
+		InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_input(msg).map_err(
+				|reason| reason.into_tx_abort_msg(self.context_mut().channel_id())),
+			None => Err(msgs::TxAbort {
+				channel_id: self.context_mut().channel_id(),
```
```suggestion
				|reason| reason.into_tx_abort_msg(self.context().channel_id())),
			None => Err(msgs::TxAbort {
				channel_id: self.context().channel_id(),
```

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+	fn tx_add_output(&mut self, msg: &msgs::TxAddOutput)-> InteractiveTxMessageSendResult {
+		InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_output(msg).map_err(
+				|reason| reason.into_tx_abort_msg(self.context_mut().channel_id())),
```
Context can be non-mutable (`context()`) in a few other places as well

dunxen replied:
Yip. Think this was a search and replace issue at some stage in a prior version.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+					ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+				)));
+		}
+		let outpoint = OutPoint { txid: signing_session.unsigned_tx.txid(), index: output_index.unwrap() };
```
It's a bit safer to use an 'if let Some(x) = x' construct with an else error branch, and avoid using `unwrap()`.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -3613,26 +3833,73 @@ pub(crate) fn per_outbound_htlc_counterparty_commit_tx_fee_msat(feerate_per_kw:
+	is_initiator: bool, funding_inputs: &[(TxIn, TransactionU16LenLimited)],
+	funding_feerate_sat_per_1000_weight: u32, holder_dust_limit_satoshis: u64,
+) -> Result<u64, APIError> {
+    use bitcoin::WScriptHash;
```
Any particular reason this `use` is local here?

dunxen replied:
Nope, just rust-analyzer being weird :)

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -3613,26 +3833,73 @@ pub(crate) fn per_outbound_htlc_counterparty_commit_tx_fee_msat(feerate_per_kw:
+			.saturating_add(TX_COMMON_FIELDS_WEIGHT)
+			// The weight of a P2WSH output to be added later.
+			.saturating_add(get_output_weight(&ScriptBuf::new_p2wsh(
+				&WScriptHash::from_slice(&[2; 32]).expect("valid scrpt hash")
```
There is a dummy constant used here, should it be explained, put in a named constant?
Also, is 'scrpt' intended or a typo?

dunxen replied:
The hash is irrelevant so I'll change it to this to avoid the `expect` and explain it better:
```rust
// The weight of a P2WSH output to be added later.
//
// NOTE: The witness script hash given here is irrelevant as it's a fixed size and we just want
// to calculate the contributed weight, so we use an all-zero hash.
.saturating_add(get_output_weight(&ScriptBuf::new_p2wsh(
	  &WScriptHash::from_raw_hash(Hash::all_zeros())
)).to_wu())
```

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4403,6 +4678,103 @@ impl<SP: Deref> Channel<SP> where
+		if self.context.commitment_secrets.get_min_seen_secret() != (1 << 48) ||
+				self.context.cur_counterparty_commitment_transaction_number != INITIAL_COMMITMENT_NUMBER ||
+				self.context.holder_commitment_point.transaction_number() != INITIAL_COMMITMENT_NUMBER {
+			panic!("Should not have advanced channel commitment tx numbers prior to funding_created");
```
How safe is this `panic` here? Could it be changed to a closing error?

dunxen replied:
This exists in V1 channels so conversation would need to happen there.

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6747,11 +6794,14 @@ where
 
-	fn do_accept_inbound_channel(&self, temporary_channel_id: &ChannelId, counterparty_node_id: &PublicKey, accept_0conf: bool, user_channel_id: u128) -> Result<(), APIError> {
-
+	// TODO(dual_funding): Remove param _-prefix once #[cfg(dual_funding)] is dropped.
```
The [cfg(dual_funding)] is dropped, isn't it?

dunxen replied:
Not all of them in this commit but in later ones :)

optout21 commented on lightning/src/ln/mod.rs:
```diff
@@ -37,6 +37,8 @@ pub mod channel;
 #[cfg(not(fuzzing))]
 pub(crate) mod channel;
 
+#[allow(dead_code)]
```
I miss a TODO here, so that it is found with higher prob. later

optout21 reviewed (changes requested):
Mostly localized code improvements suggested.

optout21 commented:
As I see, the `dual_funding` cfg flag has been removed (not a problem)

optout21 commented:
Looks to me that #2989 is in fact included in this PR (good!)

optout21 commented on lightning/src/events/mod.rs:
```diff
@@ -1176,7 +1192,7 @@ pub enum Event {
 		/// The channel value of the requested channel.
 		funding_satoshis: u64,
 		/// Our starting balance in the channel if the request is accepted, in milli-satoshi.
```
This field is explained above, but this field-comment is a bit obsolete now.

dunxen replied:
Yeah good find. It's the old `push_msat` doc. I will update it.

dunxen commented on lightning/src/events/mod.rs:
```diff
@@ -538,6 +538,17 @@ impl_writeable_tlv_based_enum!(PaymentFailureReason,
+pub enum InboundChannelFunds {
+	/// For a non-dual-funded channel, the `push_msat` value from the channel initiator to us.
+	PushMsat(u64),
+	/// Indicates the open request is for a dual funded channel.
```
It does not seem obvious so I will elaborate a bit.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7233,6 +7399,328 @@ where
 		}
 	}
 
+	fn internal_tx_add_input(&self, counterparty_node_id: &PublicKey, msg: &msgs::TxAddInput) -> Result<(), MsgHandleErrInternal> {
```
Yes I'll refactor to taking in closures

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1435,6 +1437,215 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+		let mut funding_outputs = Vec::new();
+		if self.is_initiator() {
+			let tx_out = TxOut {
+				value: Amount::from_sat(self.context().get_value_satoshis()),
```
Well at this point we know the exact amount that must go toward to channel funding output. In the holder = initiator case (which this PR does not enable yet for LDK), we would already know the acceptor's contribution (it's stated in `accept_channel` and have combined that into get_value_satoshis. Then later it is verified if the acceptor provided enough input and fees.

I do empathise with the awkwardness though and it should probably be deferred to later :+1:

dunxen reviewed (commented):
Thanks @TheBlueMatt, for review and some good points raised. I'll address these ASAP :heart:

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7617,18 +8099,46 @@ where
+
+						if chan.interactive_tx_signing_session.is_some() {
+							let monitor = try_chan_phase_entry!(self,
+								chan.commitment_signed_initial_v2(&msg, best_block, &self.signer_provider, &&logger), chan_phase_entry);
```
The previous things it implies should really be looked at in the spec. We don't want to support adding HTLCs prior to initial commitment. So I really need to investigate what is going on the spec and what might need to be fixed there.

dunxen commented:
Still working on remaining initial feedback.

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -559,6 +559,20 @@ impl_writeable_tlv_based_enum_upgradable!(PaymentFailureReason,
+	/// Indicates the open request is for a dual funded channel.
+	///
+	/// Note that these channels do not support starting with initial funds pushed from the counterparty,
+	// who is the channel opener in this case.
```
`///`

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10091,7 +10354,7 @@ where
 		// change to the contents.
 		let _persistence_guard = PersistenceNotifierGuard::optionally_notify(self, || {
-			let res = self.internal_open_channel(counterparty_node_id, msg);
+			let res = self.internal_open_channel(counterparty_node_id, OpenChannelMessage::V1(msg.clone()));
```
Could we have `OpenChannelMessage` hold a reference to avoid this clone? Or make a `OpenChannelMessageRef` type if that is needed to hold values?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7203,12 +7212,38 @@ where
+							&self.logger);
+						match channel_res {
+							Ok((channel, msg_send_event_opt)) => {
+								if let Some(msg_send_event) = msg_send_event_opt {
+									peer_state.pending_msg_events.push(msg_send_event);
+								}
```
Does the `accept_channel2` message need to be sent prior to this message?

jkczyz reviewed (commented):
Only part way through. Will continue to review.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1608,6 +1609,61 @@ pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider
+				)));
+		};
+		self.context_mut().channel_transaction_parameters.funding_outpoint = Some(outpoint);
+		let channel_transaction_parameters = self.context().channel_transaction_parameters.clone();
```
Seems like this clone should be avoidable?

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1608,6 +1609,61 @@ pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider
+		let channel_transaction_parameters = self.context().channel_transaction_parameters.clone();
+		self.context_mut().holder_signer.as_mut().provide_channel_parameters(&channel_transaction_parameters);
+
+		let commitment_signed = get_initial_commitment_signed(self.context_mut(), signing_session.unsigned_tx.clone(), logger);
```
`get_initial_commitment_signed` can take the transaction by reference to avoid this clone.

dunxen replied:
Yeah, lol

jkczyz commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -206,22 +227,162 @@ impl ConstructedTransaction {
+		if self.counterparty_tx_signatures.is_some() {
+			return (None, None);
+		};
+		self.counterparty_tx_signatures = Some(tx_signatures.clone());
```
This is cloned both here and at the call site, but only once should be fine?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7909,16 +8003,103 @@ where
+		match peer_state.channel_by_id.entry(msg.channel_id) {
+			hash_map::Entry::Occupied(mut chan_phase_entry) => {
+				let channel_phase = chan_phase_entry.get_mut();
+				let tx_abort_opt = match channel_phase {
+					ChannelPhase::UnfundedInboundV2(chan) => Self::maybe_reset_interactive_tx_state(
+						chan.interactive_tx_constructor_mut(), msg.channel_id),
+					ChannelPhase::UnfundedOutboundV2(chan) => Self::maybe_reset_interactive_tx_state(
+						chan.interactive_tx_constructor_mut(), msg.channel_id),
+					ChannelPhase::Funded(chan) => Self::maybe_reset_interactive_tx_state(
+						&mut chan.interactive_tx_constructor, msg.channel_id),
+					_ => try_chan_phase_entry!(self, Err(ChannelError::Close(
+						(
+							"Got an unexpected tx_signatures message".into(),
+							ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+						))), chan_phase_entry)
+				};
+				if let Some(tx_abort) = tx_abort_opt {
+					peer_state.pending_msg_events.push(events::MessageSendEvent::SendTxAbort {
+						node_id: *counterparty_node_id,
+						msg: tx_abort,
+					});
+				}
```
This would be more readable / cleaner if we drop the helper, IMO, since it has a side effect.

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index ac6b60c21..100141fa0 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -8052,17 +8052,6 @@ where
                }
        }
 
-       fn maybe_reset_interactive_tx_state(
-               tx_constructor: &mut Option<InteractiveTxConstructor>,
-               channel_id: ChannelId,
-       ) -> Option<msgs::TxAbort> {
-               // If we have not reset the negtotiation state yet, we must echo back a `tx_abort`.
-               tx_constructor.take().map(|_| msgs::TxAbort {
-                       channel_id,
-                       data: "Acknowledged tx_abort".to_string().into_bytes(),
-               })
-       }
-
        fn internal_tx_abort(&self, counterparty_node_id: &PublicKey, msg: &msgs::TxAbort)
        -> Result<(), MsgHandleErrInternal> {
                let per_peer_state = self.per_peer_state.read().unwrap();
@@ -8078,23 +8067,24 @@ where
                match peer_state.channel_by_id.entry(msg.channel_id) {
                        hash_map::Entry::Occupied(mut chan_phase_entry) => {
                                let channel_phase = chan_phase_entry.get_mut();
-                               let tx_abort_opt = match channel_phase {
-                                       ChannelPhase::UnfundedInboundV2(chan) => Self::maybe_reset_interactive_tx_state(
-                                               chan.interactive_tx_constructor_mut(), msg.channel_id),
-                                       ChannelPhase::UnfundedOutboundV2(chan) => Self::maybe_reset_interactive_tx_state(
-                                               chan.interactive_tx_constructor_mut(), msg.channel_id),
-                                       ChannelPhase::Funded(chan) => Self::maybe_reset_interactive_tx_state(
-                                               &mut chan.interactive_tx_constructor, msg.channel_id),
+                               let tx_constructor = match channel_phase {
+                                       ChannelPhase::UnfundedInboundV2(chan) => chan.interactive_tx_constructor_mut(),
+                                       ChannelPhase::UnfundedOutboundV2(chan) => chan.interactive_tx_constructor_mut(),
+                                       ChannelPhase::Funded(chan) => &mut chan.interactive_tx_constructor,
                                        _ => try_chan_phase_entry!(self, Err(ChannelError::Close(
                                                (
                                                        "Got an unexpected tx_signatures message".into(),
                                                        ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
                                                ))), chan_phase_entry)
                                };
-                               if let Some(tx_abort) = tx_abort_opt {
+                               if tx_constructor.take().is_some() {
+                                       let msg = msgs::TxAbort {
+                                               channel_id: msg.channel_id,
+                                               data: "Acknowledged tx_abort".to_string().into_bytes(),
+                                       };
                                        peer_state.pending_msg_events.push(events::MessageSendEvent::SendTxAbort {
                                                node_id: *counterparty_node_id,
-                                               msg: tx_abort,
+                                               msg,
                                        });
                                }
                                Ok(())
```

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1490,6 +1495,163 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+					OutputOwned::Shared(SharedOwnedOutput::new(tx_out, self.dual_funding_context().our_funding_satoshis))
+				} else {
+					OutputOwned::SharedControlFullyOwned(tx_out)
+				}
```
Yeah it is admittedly a bit confusing. I didn't have better suggestions to give at the time and the current terminology is:

owned: the holder has a stake in the value of the output which could be shared or fully owned
controlled: 
  * Full control: the holder can unilaterally spend the output and the counterparty has no spending path
  * Shared control: there must be some collaboration between the parties to spend the output

It sounds like "owned" should be named something closer to "stake" and then controlled can remain.

I believe these variants are specifically helpful for calculations and tracking in splicing.

jkczyz replied:
> Yeah it is admittedly a bit confusing. I didn't have better suggestions to give at the time and the current terminology is:
> 
> owned:
> 
> * Fully owned: the full output value belongs to the holder

What is meant by "belongs"? :)

> * Shared ownership: the holder has a stake in the value of the output
> 
> controlled:
> 
> * Full control: the holder can unilaterally spend the output and the counterparty has no spending path
> * Shared control: there must be some collaboration between the parties to spend the output
> 
> It sounds like "owned" should be named something closer to "stake" and then controlled can remain. Also if there are insufficient docs for these variants, I will update that.
> 

Yeah, expanding docs would be helpful, IMO. Also, having a concrete examples for each variant may help land on some better names.

dunxen commented:
I realise I haven't addressed https://github.com/lightningdevkit/rust-lightning/pull/3137#discussion_r1667860073 yet.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -8438,19 +8665,33 @@ impl<SP: Deref> InboundV2Channel<SP> where SP::Target: SignerProvider {
+					ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+				)));
+			}
+		};
```
Instead of returning the message to send here, I wonder if it would be cleaner if the call site directly called a method similar to `accept_inbound_dual_funded_channel` to obtain the message. That way the logic for sending the message is only applicable to the V2 case.

Currently, the assumption for V1 is that the newly created channel is always tupled with `None` at the call site. This would affect the return type of `InteractiveTxConstructor::new`, but I think that should be ok? Let me know if I'm missing something.

Also, IIUC, won't we only have a message to send if we are the initiator? (i.e., for inbound channels this will always be `None`)

dunxen replied:
Yes no message if we are acceptor. I suspect this was the result of over zealous copy pasting on my end.

dunxen replied:
I will keep the the former part of this comment in mind for outbound V2 channels.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7377,83 +7441,144 @@ where
+				let mut random_bytes = [0u8; 16];
+				random_bytes.copy_from_slice(&self.entropy_source.get_secure_random_bytes()[..16]);
+				let user_channel_id = u128::from_be_bytes(random_bytes);
+				let (mut channel, _) = match InboundV2Channel::new(&self.fee_estimator, &self.entropy_source,
```
You are correct. It will always be `None` :man_facepalming:

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -8771,6 +8999,29 @@ impl<SP: Deref> InboundV2Channel<SP> where SP::Target: SignerProvider {
+
+	pub fn funding_tx_constructed<L: Deref>(
+		mut self, mut signing_session: InteractiveTxSigningSession, logger: &L
+	) -> Result<(Channel<SP>, msgs::CommitmentSigned, Option<Event>), (Self, ChannelError)>
```
If we de-coupled the `internal_funding_tx_constructed` logic from the "convert `Self` to `Channel`" logic (i.e., where the latter is infallible), it would simplify the call sites. You wouldn't need to remove and re-insert an entry in the map for the failure case, IIUC.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -8685,11 +8685,11 @@ impl<SP: Deref> InboundV2Channel<SP> where SP::Target: SignerProvider {
+			Ok(tx_msg_opt) => {
+				// The holder is the channel acceptor and must not send the first interactive transaction message.
+				debug_assert!(tx_msg_opt.is_none());
+			},
```
That would be simpler to reason about and not hide anything, plus get rid of these checks. Will update.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -4480,6 +4759,105 @@ impl<SP: Deref> Channel<SP> where
 		Ok(())
 	}
 
+	pub fn commitment_signed_initial_v2<L: Deref>(
```
Here we are just adding it for inbound V2 channels, but actually, eventually outbound will also need it once that's enabled (https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#the-commitment_signed-message), and I'll probably share the behaviour via `InteractivelyFunded` trait or something like that.

dunxen replied:
Ok so after looking at this for a while, it's something we can share behaviour for via the template method pattern Jeff suggested and have on `OutboundV2Channel` impl to align with `funding_signed` in `OutboundV1Channel` and similarly for `InboundV2Channel/InboundV1Channel`. However, this is under the assumption that there is a similar ordering to the sending of the initial `commitment_signed`.

For V1, exchange of commitment sigs is always strictly ordered:
Initiator followed by acceptor as `funding_signed` strictly follows `funding_created`.

For V2, there is no prescribed order of sending `commitment_signed`s, so the initiator or acceptor could technically send theirs first.

As acceptor we would by no means violate the spec if we just assume the channel initiator will send `commitment_signed` first and then we follow after that and then promote `InboundV2Channel` to a `Channel`. This would be like acceptor getting `funding_created` in the V1 case.

If we are the initiator, from our point of view immediately following consecutive `tx_complete` msgs, we queue a `commitment_signed` and promote `OutboundV2Channel` to a `Channel`.

With those reasonable assumptions, we can keep things in V2 establishment analogous to V1 for initial commitment signature exchange?

dunxen replied:
Fair enough around no new assumptions and this can definitely be DRYed up in a prefactor commit!

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -8375,34 +8807,66 @@ impl<SP: Deref> OutboundV2Channel<SP> where SP::Target: SignerProvider {
+		let channel = Channel {
+			context: self.context,
+			dual_funding_channel_context: Some(self.dual_funding_context),
+			interactive_tx_constructor: self.interactive_tx_constructor,
```
You're right, I initially thought we might need to keep around the inputs/outputs if the initiator requests an RBF, but that's actually not necessary. We will require the user to respond to a fee bump request via some event anyway. I also think we defer any thinking around that to the #3281 design.

We will still have everything necessary for the `next_funding` tlv in the channel reestablishment and that's all we need at the moment: https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#message-retransmission

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7903,28 +7952,150 @@ where
+									update_fee: None,
+								},
+							});
+						}).map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))
```
When possible, it's best to avoid side-effects inside `map` and `and_then` calls. It would make the code easier to follow and less indented. Consider the following:

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index b23bd08ff..c2e67287a 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -7955,7 +7955,7 @@ where
                                };
                                if let Some(mut signing_session) = signing_session_opt {
                                        let funding_txid = signing_session.unsigned_tx.txid();
-                                       match chan_phase_entry.get_mut() {
+                                       let (commitment_signed, funding_ready_for_sig_event_opt) = match chan_phase_entry.get_mut() {
                                                ChannelPhase::UnfundedOutboundV2(chan) => {
                                                        chan.funding_tx_constructed(&mut signing_session, &self.logger)
                                                },
@@ -7965,45 +7965,39 @@ where
                                                _ => Err(ChannelError::Warn(
                                                        "Got a tx_complete message with no interactive transaction construction expected or in-progress"
                                                        .into())),
-                                       }.and_then(|(commitment_signed, funding_ready_for_sig_event_opt)| {
-                                                       let (channel_id, channel_phase) = chan_phase_entry.remove_entry();
-                                                       match channel_phase {
-                                                               ChannelPhase::UnfundedOutboundV2(chan) => {
-                                                                       chan.into_channel(signing_session)
-                                                               },
-                                                               ChannelPhase::UnfundedInboundV2(chan) => {
-                                                                       chan.into_channel(signing_session)
-                                                               },
-                                                               _ => {
-                                                                       debug_assert!(false); // It cannot be another variant as we are in the `Ok` branch of the above match.
-                                                                       Err(ChannelError::Warn(
-                                                                               "Got a tx_complete message with no interactive transaction construction expected or in-progress"
-                                                                                       .into()))
-                                                               }
-                                                       }.map(|channel| (channel_id, channel, funding_ready_for_sig_event_opt, commitment_signed))
-                                               }).map(|(channel_id, mut channel, funding_ready_for_sig_event_opt, commitment_signed)| {
-                                                       channel.set_next_funding_txid(&funding_txid);
-                                                       peer_state.channel_by_id.insert(channel_id, ChannelPhase::Funded(channel));
-                                                       if let Some(funding_ready_for_sig_event) = funding_ready_for_sig_event_opt {
-                                                               let mut pending_events = self.pending_events.lock().unwrap();
-                                                               pending_events.push_back((funding_ready_for_sig_event, None));
-                                                       }
-                                                       peer_state.pending_msg_events.push(events::MessageSendEvent::UpdateHTLCs {
-                                                               node_id: counterparty_node_id,
-                                                               updates: CommitmentUpdate {
-                                                                       commitment_signed,
-                                                                       update_add_htlcs: vec![],
-                                                                       update_fulfill_htlcs: vec![],
-                                                                       update_fail_htlcs: vec![],
-                                                                       update_fail_malformed_htlcs: vec![],
-                                                                       update_fee: None,
-                                                               },
-                                                       });
-                                               }).map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))
-                               } else {
-                                       // We're not in a signing session yet so we don't need to do anything else.
-                                       Ok(())
+                                       }.map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))?;
+
+                                       let (channel_id, channel_phase) = chan_phase_entry.remove_entry();
+                                       let mut channel = match channel_phase {
+                                               ChannelPhase::UnfundedOutboundV2(chan) => chan.into_channel(signing_session),
+                                               ChannelPhase::UnfundedInboundV2(chan) => chan.into_channel(signing_session),
+                                               _ => {
+                                                       debug_assert!(false); // It cannot be another variant as we are in the `Ok` branch of the above match.
+                                                       Err(ChannelError::Warn(
+                                                               "Got a tx_complete message with no interactive transaction construction expected or in-progress"
+                                                                       .into()))
+                                               },
+                                       }.map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))?;
+
+                                       channel.set_next_funding_txid(&funding_txid);
+                                       peer_state.channel_by_id.insert(channel_id, ChannelPhase::Funded(channel));
+                                       if let Some(funding_ready_for_sig_event) = funding_ready_for_sig_event_opt {
+                                               let mut pending_events = self.pending_events.lock().unwrap();
+                                               pending_events.push_back((funding_ready_for_sig_event, None));
+                                       }
+                                       peer_state.pending_msg_events.push(events::MessageSendEvent::UpdateHTLCs {
+                                               node_id: counterparty_node_id,
+                                               updates: CommitmentUpdate {
+                                                       commitment_signed,
+                                                       update_add_htlcs: vec![],
+                                                       update_fulfill_htlcs: vec![],
+                                                       update_fail_htlcs: vec![],
+                                                       update_fail_malformed_htlcs: vec![],
+                                                       update_fee: None,
+                                               },
+                                       });
                                }
+                               Ok(())
                        },
                        hash_map::Entry::Vacant(_) => {
                                Err(MsgHandleErrInternal::send_err_msg_no_close(format!("Got a message for a channel from the wrong node! No such channel for the passed counterparty_node_id {}", counterparty_node_id), msg.channel_id))
```

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -9530,7 +9542,8 @@ impl<SP: Deref> Writeable for Channel<SP> where SP::Target: SignerProvider {
 			(51, is_manual_broadcast, option), // Added in 0.0.124
-			(53, funding_tx_broadcast_safe_event_emitted, option) // Added in 0.0.124
+			(53, funding_tx_broadcast_safe_event_emitted, option), // Added in 0.0.124
+			(55, self.context.next_funding_txid, option) // Added in 0.0.125
```
Need to add this to the `ReadableArgs` implementation as discussed offline.

optout21 commented on lightning/src/ln/mod.rs:
```diff
@@ -96,7 +96,7 @@ mod async_signer_tests;
 #[allow(unused_mut)]
 mod offers_tests;
-#[allow(dead_code)] // TODO(dual_funding): Exchange for dual_funding cfg
+#[allow(dead_code)] // TODO(splicing): Exchange for splicing cfg
```
I think the dead_code exception no longer needed. This change seems to be search/and/replace false positive.

dunxen replied:
Just checked. Is still needed for now :(

dunxen commented:
> Needs rebase now :(

Sorry, catching up on everything now! Been mostly offline for a week.

dunxen commented:
Just rebased for now while I'm still addressing past week's feedback.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -8756,6 +9023,85 @@ fn get_initial_channel_type(config: &UserConfig, their_features: &InitFeatures)
 	ret
 }
 
+fn get_initial_counterparty_commitment_signature<SP:Deref, L: Deref>(
```
Definitely makes sense now.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4801,6 +4804,53 @@ where
 		result
 	}
 
+	/// Handles a signed funding transaction generated by interactive transaction construction and
+	/// provided by the client.
```
This isn't actually relevant to this PR yet as we don't contribute any inputs/funds, but I've made a not for updating the docs when we do. Removing the method and anything related to signing contributed inputs for now.

dunxen commented:
DRY'd up new initial commitment in latest push. Now moving over some tests from #2302 and modifying them to have manual V2 open channel messages so we can at least test them here.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1597,7 +1597,59 @@ impl<SP: Deref> InteractivelyFunded<SP> for InboundV2Channel<SP> where SP::Targe
+	fn context(&self) -> &ChannelContext<SP>;
+
+	fn initial_commitment_signed<L: Deref>(
+		&mut self, holder_commitment_tx: HolderCommitmentTransaction,
```
Ok yeah I was a bit too prudent in sharing the code but there are still some obvious pieces like you pointed out. Will push them up in the existing DRY commit and a new commit with tests.

dunxen commented:
Pushed the DRYing up. Migrating the appropriate tests from 2302 is taking some time as more manual helpers methods are needed. Almost there, though.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7922,6 +8027,265 @@ where
+					ChannelPhase::Funded(chan) => &mut chan.interactive_tx_constructor,
+					_ => try_chan_phase_entry!(self, Err(ChannelError::Close(
+						(
+							"Got an unexpected tx_signatures message".into(),
```
Should this say `tx_abort` message?

jkczyz reviewed (commented):
Some comments based on our offline discussion. Will take a closer look at your recent changes.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3676,33 +4038,81 @@ pub(crate) fn get_legacy_default_holder_selected_channel_reserve_satoshis(channe
+	#[allow(dead_code)] // TODO(dual_funding): Remove once contribution to V2 channels is enabled.
 	pub dual_funding_channel_context: Option<DualFundingChannelContext>,
+	/// The current interactive transaction construction session under negotiation.
+	pub interactive_tx_constructor: Option<InteractiveTxConstructor>,
+	pub interactive_tx_signing_session: Option<InteractiveTxSigningSession>,
```
Sorry, that should have said `InteractiveTxConstructor` when in `ChannelPhase::Funded`.

dunxen replied:
For now, it won't be used so I'll remove it from `Channel`. We can bring it up again when addressing RBF, because that's where it may be needed depending on the design.

dunxen commented:
> Git history is weird here looks like some commits from upstream got cherry-picked, can just rebase to fix it.

Yeah, I see now. Fixing.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1598,7 +1598,152 @@ impl<SP: Deref> InteractivelyFunded<SP> for InboundV2Channel<SP> where SP::Targe
-impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider  {
+/// A channel struct implementing this trait can receive an initial counterparty commitment
+/// transaction signature.
+trait InitialRemoteCommitmentReceiver<SP: Deref> where SP::Target: SignerProvider {
```
#3370

dunxen commented:
Now depends on #3370, so have rebased on that and will continue to address outstanding feedback.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -8440,18 +8838,45 @@ impl<SP: Deref> InboundV2Channel<SP> where SP::Target: SignerProvider {
+			    expected_remote_shared_funding_output: Some((channel.context().get_funding_redeemscript(), channel.context().channel_value_satoshis)),
+			}
+		) {
+			Ok(tx_constructor) => {
+				channel.set_interactive_tx_constructor(tx_constructor);
+			},
```
Oh yeah, duh! Thanks! :)

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -4093,9 +4093,6 @@ pub(super) struct DualFundingChannelContext {
 	pub context: ChannelContext<SP>,
-	pub dual_funding_channel_context: Option<DualFundingChannelContext>,
-	/// The current interactive transaction construction session under negotiation.
-	pub interactive_tx_constructor: Option<InteractiveTxConstructor>,
```
This is added in an earlier commit than the commit this fixup is applied to.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -4093,9 +4093,6 @@ pub(super) struct DualFundingChannelContext {
 // Counterparty designates channel data owned by the another channel participant entity.
 pub(super) struct Channel<SP: Deref> where SP::Target: SignerProvider {
 	pub context: ChannelContext<SP>,
-	pub dual_funding_channel_context: Option<DualFundingChannelContext>,
```
Looks like this was added prior to this PR. Let's remove it in an independent commit rather than a fixup so that it is clear why it was removed.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7626,83 +7687,144 @@ where
+				temporary_channel_id.clone()));
+		}
+
+		// Version-specific checks and logic
```
Looks like we are only accessing `msgs.common_fields`, so we can get a reference to that at the beginning of the method, if needed.

dunxen commented:
> Ah looks like the last round of comments wasn't fully addressed yet either, sorry got excited :)

Ah sorry. I didn't actually send the comment. Addressing your feedback this morning.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1642,6 +1647,172 @@ impl<SP: Deref> InitialRemoteCommitmentReceiver<SP> for InboundV1Channel<SP> whe
+					ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+				)));
+		};
+		self.context_mut().channel_transaction_parameters.funding_outpoint = Some(outpoint);
```
I'll put it in this PR or create an issue once I'm done with other feedback here.

dunxen replied:
https://github.com/lightningdevkit/rust-lightning/issues/3392

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -5138,6 +5604,58 @@ impl<SP: Deref> Channel<SP> where
+
+				// TODO(dual_funding): Check all sigs are SIGHASH_ALL.
+
+				// TODO(dual_funding): I don't see how we're going to be able to ensure witness-standardness
```
Thanks, will document and add release notes in this PR.

dunxen commented:
Another round of fix-ups. The review comments not marked as resolved, I'm still working on.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7922,6 +8027,265 @@ where
+						msg,
+					});
+				}
+				Ok(())
```
`tx_abort` Just aborts the interactive construction, in principle it could be started again with initiator sending `tx_add_input` as [per the spec](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#the-tx_abort-message).

I can document this here too and add a link.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7694,116 +7694,75 @@ where
 						return Err(MsgHandleErrInternal::from_chan_no_close(e, msg.common_fields.temporary_channel_id));
 					},
 					Ok(res) => res
-				};
```
Let's use `map_err()?` instead of inlining a `match`. Same for V2.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7812,26 +7771,38 @@ where
+		}
+		if channel_type.requires_anchors_zero_fee_htlc_tx() {
+			return Err(MsgHandleErrInternal::send_err_msg_no_close("No channels with anchor outputs accepted".to_owned(), common_fields.temporary_channel_id));
+		}
```
Predates this PR, but I wonder why we don't do this earlier using `channel::channel_type_from_open_channel`. If we can, then it would make the rest of the code simpler since the `MessageSendEvent` can be push directly in the `match` against `msg` and `outbound_scid_alias` can be passed directly to the constructors.

dunxen replied:
Yeah thanks, moving it above the `manually_accept_inbound_channels` check works just fine. We need it immediately in both cases either way. The zero conf and anchors checks I have left just after the manual acceptance check.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7812,26 +7771,38 @@ where
+				node_id: *counterparty_node_id,
+				msg: channel.accept_inbound_dual_funded_channel(),
+			},
+			_ => {
+				debug_assert!(false, "We can only have an unfunded inbound V1/V2 channel here");
+				return Err(MsgHandleErrInternal::send_err_msg_no_close("This should be an inbound channel.".to_owned(), common_fields.temporary_channel_id))
+			}
```
We can get rid of this branch with the following rearrangement. Though note that the `MessageSendEvent` is created prior to the checks made against the channel type that are mentioned in another comment.

```diff
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 20810b000..4e9bcb6e2 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -8504,7 +8504,7 @@ impl<SP: Deref> InboundV1Channel<SP> where SP::Target: SignerProvider {
        /// should be sent back to the counterparty node.
        ///
        /// [`msgs::AcceptChannel`]: crate::ln::msgs::AcceptChannel
-       pub fn accept_inbound_channel(&mut self) -> msgs::AcceptChannel {
+       pub fn accept_inbound_channel(&self) -> msgs::AcceptChannel {
                if self.context.is_outbound() {
                        panic!("Tried to send accept_channel for an outbound channel?");
                }
@@ -8881,7 +8881,7 @@ impl<SP: Deref> InboundV2Channel<SP> where SP::Target: SignerProvider {
        /// should be sent back to the counterparty node.
        ///
        /// [`msgs::AcceptChannelV2`]: crate::ln::msgs::AcceptChannelV2
-       pub fn accept_inbound_dual_funded_channel(&mut self) -> msgs::AcceptChannelV2 {
+       pub fn accept_inbound_dual_funded_channel(&self) -> msgs::AcceptChannelV2 {
                if self.context.is_outbound() {
                        debug_assert!(false, "Tried to send accept_channel for an outbound channel?");
                }
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 55ac16802..7c3f10f4c 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -7767,30 +7767,32 @@ where
                random_bytes.copy_from_slice(&self.entropy_source.get_secure_random_bytes()[..16]);
                let user_channel_id = u128::from_be_bytes(random_bytes);
 
-               let mut channel_phase = match msg {
+               let (mut channel_phase, message_send_event) = match msg {
                        OpenChannelMessageRef::V1(msg) => {
-                               ChannelPhase::UnfundedInboundV1(match InboundV1Channel::new(
+                               let channel = InboundV1Channel::new(
                                        &self.fee_estimator, &self.entropy_source, &self.signer_provider, *counterparty_node_id,
                                        &self.channel_type_features(), &peer_state.latest_features, msg, user_channel_id,
                                        &self.default_configuration, best_block_height, &self.logger, /*is_0conf=*/false
-                               ) {
-                                       Err(e) => {
-                                               return Err(MsgHandleErrInternal::from_chan_no_close(e, msg.common_fields.temporary_channel_id));
-                                       },
-                                       Ok(res) => res
-                               })
+                               )
+                                       .map_err(|e| MsgHandleErrInternal::from_chan_no_close(e, msg.common_fields.temporary_channel_id))?;
+                               let message_send_event = events::MessageSendEvent::SendAcceptChannel {
+                                       node_id: *counterparty_node_id,
+                                       msg: channel.accept_inbound_channel(),
+                               };
+                               (ChannelPhase::UnfundedInboundV1(channel), message_send_event)
                        },
                        OpenChannelMessageRef::V2(msg) => {
-                               ChannelPhase::UnfundedInboundV2(match InboundV2Channel::new(&self.fee_estimator, &self.entropy_source,
-                                       &self.signer_provider, self.get_our_node_id(), *counterparty_node_id, &self.channel_type_features(),
-                                       &peer_state.latest_features, msg, vec![], user_channel_id, &self.default_configuration,
-                                       best_block_height, &self.logger)
-                               {
-                                       Err(e) => {
-                                               return Err(MsgHandleErrInternal::from_chan_no_close(e, msg.common_fields.temporary_channel_id));
-                                       },
-                                       Ok(res) => res
-                               })
+                               let channel = InboundV2Channel::new(
+                                       &self.fee_estimator, &self.entropy_source, &self.signer_provider, self.get_our_node_id(),
+                                       *counterparty_node_id, &self.channel_type_features(), &peer_state.latest_features, msg,
+                                       vec![], user_channel_id, &self.default_configuration, best_block_height, &self.logger
+                               )
+                                       .map_err(|e| MsgHandleErrInternal::from_chan_no_close(e, msg.common_fields.temporary_channel_id))?;
+                               let message_send_event = events::MessageSendEvent::SendAcceptChannelV2 {
+                                       node_id: *counterparty_node_id,
+                                       msg: channel.accept_inbound_dual_funded_channel(),
+                               };
+                               (ChannelPhase::UnfundedInboundV2(channel), message_send_event)
                        },
                };
 
@@ -7805,20 +7807,6 @@ where
                let outbound_scid_alias = self.create_and_insert_outbound_scid_alias();
                channel_phase.context_mut().set_outbound_scid_alias(outbound_scid_alias);
 
-               let message_send_event = match channel_phase {
-                       ChannelPhase::UnfundedInboundV1(ref mut channel) => events::MessageSendEvent::SendAcceptChannel {
-                               node_id: *counterparty_node_id,
-                               msg: channel.accept_inbound_channel(),
-                       },
-                       ChannelPhase::UnfundedInboundV2(ref mut channel) => events::MessageSendEvent::SendAcceptChannelV2 {
-                               node_id: *counterparty_node_id,
-                               msg: channel.accept_inbound_dual_funded_channel(),
-                       },
-                       _ => {
-                               debug_assert!(false, "We can only have an unfunded inbound V1/V2 channel here");
-                               return Err(MsgHandleErrInternal::send_err_msg_no_close("This should be an inbound channel.".to_owned(), common_fields.temporary_channel_id))
-                       }
-               };
                peer_state.pending_msg_events.push(message_send_event);
                peer_state.channel_by_id.insert(channel_phase.context().channel_id(), channel_phase);
```

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7533,14 +7576,31 @@ where
+						peer_state.channel_by_id.insert(channel.context.channel_id(),
+							ChannelPhase::UnfundedInboundV2(channel));
+					},
+					_ => {
+						debug_assert!(false);
+						// This should be unreachable, but if it is then we would have dropped the inbound channel request
+						// and there'd be nothing to clean up as we haven't added anything to the channel_by_id map yet.
+						return Err(APIError::APIMisuseError {
+							err: "Channel somehow changed to a non-inbound channel before accepting".to_owned() })
+					}
```
Probably can do a similar changes here as commented on in `internal_open_channel` to avoid this branch.

dunxen commented:
Thanks for the feedback, @jkczyz. Addressing it today.

jkczyz reviewed (commented):
Mostly minor things while going through a last pass.

dunxen commented:
@TheBlueMatt this ready for another pass when you get a gap :pray:

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -8581,6 +8999,59 @@ fn get_initial_channel_type(config: &UserConfig, their_features: &InitFeatures)
+	let signature = match context.get_initial_counterparty_commitment_signature(logger) {
+		Ok(res) => res,
+		Err(e) => {
+			log_error!(logger, "Got bad signatures: {:?}!", e);
```
Yeah this is ancient when pulled over from get_funding_created some time last year in #2302. I think it may be a good idea to just add async signer support in this PR and not leave that for down the road.

dunxen replied:
https://github.com/lightningdevkit/rust-lightning/issues/3404

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -5298,6 +5625,58 @@ impl<SP: Deref> Channel<SP> where
 	}
 
+	pub fn tx_signatures(&mut self, msg: &msgs::TxSignatures) -> Result<(Option<msgs::TxSignatures>, Option<Transaction>), ChannelError> {
+		if let Some(ref mut signing_session) = self.interactive_tx_signing_session {
```
We should check state and wait for persistence of initial `ChannelMonitor` here, but in `InteractiveTxSigningSession` we will only allow our `tx_signatures` to be sent if we have received a `commitment_signed`, but looks like something went missing in `commitment_signed_initial_v2` along the way.

Nothing broke in the test as we always send an empty witnesses array in `tx_signatures` since we don't support contributing funds/inputs here. I'll make sure it goes through the full flow in this PR anyway so we can catch that error.

Thanks!

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -8581,6 +8999,59 @@ fn get_initial_channel_type(config: &UserConfig, their_features: &InitFeatures)
 	ret
 }
 
+fn get_initial_commitment_signed<SP:Deref, L: Deref>(
```
Yes!

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1494,6 +1495,21 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+	// transaction construction, or safely abort that transaction if it was not signed by one of the
+	// peers, who has thus already removed it from its state.
+	// 
+	// If we've sent `commtiment_signed` for an interactive an interactively constructed transaction
```
oops
```suggestion
	// If we've sent `commtiment_signed` for an interactively constructed transaction
```

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1494,6 +1495,21 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+	//
+	// TODO(dual_funding): Persist this when we actually contribute funding inputs. For now we always
+	// send an empty witnesses array in `tx_signatures` as a V2 channel acceptor
+	next_funding_txid: Option<Txid>,
```
Yeah that's better to reuse for persistence and I think we should be populating that when sending our initial `commitment_signed` I see. Will update it in this PR already so it plays nice.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -3822,33 +4070,77 @@ pub(crate) fn get_legacy_default_holder_selected_channel_reserve_satoshis(channe
+	#[allow(dead_code)] // TODO(dual_funding): Remove once contribution to V2 channels is enabled.
 	/// The amount in satoshis our counterparty will be contributing to the channel.
-	pub their_funding_satoshis: u64,
+	pub their_funding_satoshis: Option<u64>,
```
Yeah let me remove this change for now :)

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3969,56 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+			ChannelSignerType::Ecdsa(ref ecdsa) => {
+				ecdsa.sign_counterparty_commitment(&counterparty_initial_commitment_tx, Vec::new(), Vec::new(), &self.secp_ctx)
+					.map(|(signature, _)| signature)
+					.map_err(|_| ChannelError::Close(
```
I think it's best to get parity in this PR. I don't believe it will be that much extra work and we can DRY up as you said.

https://github.com/lightningdevkit/rust-lightning/pull/3137/files#r1831690110

dunxen replied:
Alright I'll leave this until I've fixed everything else here and then decide.

dunxen replied:
https://github.com/lightningdevkit/rust-lightning/issues/3404

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -3822,33 +4070,77 @@ pub(crate) fn get_legacy_default_holder_selected_channel_reserve_satoshis(channe
+	for (idx, input) in funding_inputs.iter().enumerate() {
+		if let Some(output) = input.1.as_transaction().output.get(input.0.previous_output.vout as usize) {
+			total_input_satoshis = total_input_satoshis.saturating_add(output.value.to_sat());
+			our_contributed_weight = our_contributed_weight.saturating_add(estimate_input_weight(output).to_wu());
```
Sorry for missing this comment which is a super important one :man_facepalming: 

It may be possible if they provide "dummy" witnesses or provide the sizes for the witnesses.

They won't be able to provide valid witnesses (that involve signatures) before construction of the funding transaction is finished.

The issue is since this transaction is always using SIGHASH_ALL for any inputs sigs, we don't know what the sighash we need to sign will be for the funding tx at this point as we have no idea of the inputs or outputs the counterparty will provide (except for the 2-of-2 funding output of course). So the signatures cannot be created by the user before construction is done. At best we can estimate based on the output type for each input provided by the user. 

Thoughts? Or has Fry-day fried my brain? I'm actually hoping that's the case and it's a super trivial fix :P

dunxen replied:
Yeah we don't care about the weight of the counterparty's inputs/outputs, and we could ask the user to provide their own witness weights or dummy witnesses that use dummy signatures where appropriate.

What they won't be able to provide is *valid* signatures in witnesses until after the whole tx is constructed as it's SIGHASH_ALL, but we don't need them yet, just the sizes.

We will have to ask the user again to provide those valid witnesses after transaction construction like we do are doing for V2.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1726,8 +1727,13 @@ pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider
 		HandleTxCompleteResult(match self.interactive_tx_constructor_mut() {
-			Some(ref mut tx_constructor) => tx_constructor.handle_tx_complete(msg).map_err(
-				|reason| reason.into_tx_abort_msg(self.context().channel_id())),
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_complete(msg)
+				.inspect(|tx_complete_value| {
+					if let HandleTxCompleteValue::SendTxComplete(_, signing_session) = tx_complete_value {
+						self.context_mut().next_funding_txid = Some(signing_session.unsigned_tx.txid());
+					}
+				})
```
Can't used `inspect` as it hasn't made it into our MSRV (1.63.0).

dunxen replied:
Yeah I've removed it locally and haven't pushed latest changes yet :+1:

jkczyz commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -1062,6 +1320,39 @@ pub(crate) enum InteractiveTxMessageSend {
 	TxComplete(msgs::TxComplete),
 }
 
+impl InteractiveTxMessageSend {
+	pub fn into_msg_send_event(self, counterparty_node_id: PublicKey) -> MessageSendEvent {
+		match self {
+			InteractiveTxMessageSend::TxAddInput(msg) => {
+				MessageSendEvent::SendTxAddInput { node_id: counterparty_node_id, msg }
+			},
+			InteractiveTxMessageSend::TxAddOutput(msg) => {
+				MessageSendEvent::SendTxAddOutput { node_id: counterparty_node_id, msg }
+			},
+			InteractiveTxMessageSend::TxComplete(msg) => {
+				MessageSendEvent::SendTxComplete { node_id: counterparty_node_id, msg }
+			},
+		}
+	}
+}
+
+pub(super) struct InteractiveTxMessageSendResult(
+	pub Result<InteractiveTxMessageSend, msgs::TxAbort>,
+);
+
+impl InteractiveTxMessageSendResult {
+	pub fn into_msg_send_event(self, counterparty_node_id: PublicKey) -> MessageSendEvent {
+		match self.0 {
+			Ok(interactive_tx_msg_send) => {
+				interactive_tx_msg_send.into_msg_send_event(counterparty_node_id)
+			},
+			Err(tx_abort_msg) => {
+				MessageSendEvent::SendTxAbort { node_id: counterparty_node_id, msg: tx_abort_msg }
+			},
+		}
+	}
+}
```
Shouldn't hold up this PR, but is there any reason why we don't simply add a `TxAbort` variant to `InteractiveTxMessageSend`? Then we wouldn't need either `InteractiveTxMessageSendResult` or `HandleTxCompleteResult`, IIUC, which would remove two layers of structs.

Otherwise, we go through the following conversions just to end up with a `MessageSendEvent`:
- `Result<InteractiveTxMessageSend, AbortReason>`
- `InteractiveTxMessageSendResult(Result<InteractiveTxMessageSend, msgs::TxAbort>)`
- `MessageSendEvent`

The `Result`s don't seem to buy us anything more than the extra variant would as we aren't doing anything special for the `Err` case. Then it would simply be:
- `InteractiveTxMessageSend`
- `MessageSendEvent`

And AFAICT, `InteractiveTxConstructor` has the `channel_id`, so there's no need to grab it from the `ChannelContext`. Or will it change in some case?

dunxen replied:
Yes, this would work. I imagine the original iteration I had in mind has some `Err` case to handle which is not applicable anymore.

Yeah the channel_id is fixed once we have a constructor. I'm quite sure the same is true of splicing, but it can be refactored if that's not the case.

dunxen replied:
I'm going to leave this as is for now and just apply your other feedback. `TxAbort` is currently handled distinctly from `InteractiveTxMessageSend`. It is handled at the `Channel` level. I suspect we can have it handled that same as any other variant of `InteratveTxMessageSend` without issues but it would require some refactoring. Will create a tracking issue and leave for now.

jkczyz reviewed (commented):
Mostly looks good. Let me know if there's anything else outstanding that needs to be addressed still.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1642,6 +1661,195 @@ impl<SP: Deref> InitialRemoteCommitmentReceiver<SP> for InboundV1Channel<SP> whe
+
+	fn tx_complete(&mut self, msg: &msgs::TxComplete) -> HandleTxCompleteResult {
+		HandleTxCompleteResult(match self.interactive_tx_constructor_mut() {
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_complete(msg)
+				// TODO: Change this to Result::inspect when MSRV >= 1.76
+				// See: https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.inspect
+				.map(|tx_complete_value| {
+					if let HandleTxCompleteValue::SendTxComplete(_, ref signing_session) = tx_complete_value {
+						self.context_mut().next_funding_txid = Some(signing_session.unsigned_tx.txid());
+					}
+					tx_complete_value
+				})
+				.map_err(|reason| reason.into_tx_abort_msg(self.context().channel_id())),
```
Agreed. Will move the side effect as suggested.

dunxen commented:
> Mostly looks good. Let me know if there's anything else outstanding that needs to be addressed still.

Async persist test mentioned in https://github.com/lightningdevkit/rust-lightning/pull/3137#discussion_r1832969732 still needs to be added.

jkczyz reviewed (commented):
Just working on async signing and was wondering about a couple things.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1642,6 +1661,202 @@ impl<SP: Deref> InitialRemoteCommitmentReceiver<SP> for InboundV1Channel<SP> whe
+					ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+				)));
+		};
+		context.channel_transaction_parameters.funding_outpoint = Some(outpoint);
+		context.holder_signer.as_mut().provide_channel_parameters(&context.channel_transaction_parameters);
```
These are overridden in `get_initial_commitment_signed`, which is called next. Do they need to be set here?

dunxen replied:
Will remove the duplication in `get_initial_commitment_signed`

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3991,109 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+			}
+		};
+
+		if self.signer_pending_funding {
+			log_trace!(logger, "Counterparty commitment signature ready for funding_created message: clearing signer_pending_funding");
+			self.signer_pending_funding = false;
+		}
```
As discussed offline, this can be dropped from this PR and re-added for async signing.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3991,95 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+			Ok(res) => res,
+			Err(e) => {
+				log_error!(logger, "Got bad signatures: {:?}!", e);
+				self.channel_transaction_parameters.funding_outpoint = None;
```
Could you clear this at the call site when handling the error? It's set prior to calling `get_initial_commitment_signed`, so clearing it there would lead to less obscurity.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3991,95 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+			if flags == (NegotiatingFundingFlags::OUR_INIT_SENT | NegotiatingFundingFlags::THEIR_INIT_SENT)) {
+			panic!("Tried to get a funding_created messsage at a time other than immediately after initial handshake completion (or tried to get funding_created twice)");
+		}
+		if self.commitment_secrets.get_min_seen_secret() != (1 << 48) ||
+				self.cur_counterparty_commitment_transaction_number != INITIAL_COMMITMENT_NUMBER ||
+				self.holder_commitment_point.transaction_number() != INITIAL_COMMITMENT_NUMBER {
+			panic!("Should not have advanced channel commitment tx numbers prior to initial commitment_signed");
+		}
```
Should this use `assert_no_commitment_advancement` instead? Or do we need the `panic` instead of `debug_assert`?

jkczyz replied:
Although, I guess the message used in `assert_no_commitment_advancement` is different.

dunxen replied:
Fixed by allowing to take a message

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3991,95 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+		if !matches!(
+			self.channel_state, ChannelState::NegotiatingFunding(flags)
+			if flags == (NegotiatingFundingFlags::OUR_INIT_SENT | NegotiatingFundingFlags::THEIR_INIT_SENT)) {
+			panic!("Tried to get a funding_created messsage at a time other than immediately after initial handshake completion (or tried to get funding_created twice)");
```
s/funding_created/initial commitment_signed ?

jkczyz commented on lightning/src/ln/dual_funding_tests.rs:
```diff
@@ -0,0 +1,266 @@
+		V2ChannelEstablishmentTestSession { initiator_input_value_satoshis: 100_000 },
+		false,
+	);
+	// Only initiator contributes, no pending
```
s/no pending/persist pending

dunxen replied:
thanks. that was a botched comment of note lol

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11702,9 +11863,13 @@ where
-			"Dual-funded channels not supported".to_owned(),
-			msg.channel_id.clone())), counterparty_node_id);
+		// Note that we never need to persist the updated ChannelManager for an inbound
+		// tx_add_input message - interactive transaction construction does not need to
```
s/tx_add_input/tx_abort

dunxen replied:
Ah missed these. Thanks

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11431,39 +11817,58 @@ where
-			"Dual-funded channels not supported".to_owned(),
-			msg.channel_id.clone())), counterparty_node_id);
+		// Note that we never need to persist the updated ChannelManager for an inbound
+		// tx_add_input message - interactive transaction construction does not need to
```
Looks like the message names in the comments need to be updated here and below.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1642,6 +1661,205 @@ impl<SP: Deref> InitialRemoteCommitmentReceiver<SP> for InboundV1Channel<SP> whe
+		HandleTxCompleteResult(Ok(tx_complete))
+	}
+
+	fn funding_tx_constructed<L: Deref>(
```
Just note that we'll still want this as a function since it's called by `signer_unblocked` for async signing.

jkczyz replied:
See #3411

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -1642,6 +1661,205 @@ impl<SP: Deref> InitialRemoteCommitmentReceiver<SP> for InboundV1Channel<SP> whe
+	fn tx_add_input(&mut self, msg: &msgs::TxAddInput) -> InteractiveTxMessageSendResult {
+		InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
+			Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_input(msg).map_err(
+				|reason| reason.into_tx_abort_msg(self.context().channel_id())),
```
It should not. It should only cancel the negotiation. 

https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#rationale-7 :

> The tx_abort message allows for the cancellation of an in progress negotiation, and a return to the initial starting state. It is distinct from the error message, which triggers a channel close.

dunxen replied:
The initiator should be able to restart negotiation with a `tx_add_input` message again.

dunxen commented:
> Few more comments, got through the bulk of it. Basically the only issue(s) I think need fixing are the persistence of `Channel`s before we have a `ChannelMonitor` which is gonna cause issues on restart. It can be fixed in a followup, though. Still need to review the tests but I can do that after this lands.

Thanks! Going to address a few more things you brought up throughout tonight and tomorrow morning for me and then I'll open an issue with remaining follow-ups. So unless something major prevents this from landing, we could land it tomorrow?

dunxen commented:
I've pushed up a few fixes. There are some nits that could be fixed here, but I'll include them in a followup to start unblocking other PRs.

@jkczyz, if CI and you are happy, we can go ahead and get this in when you're ready. Then I'll create the follow-up issue and tag it to make sure it's a blocker for next release.

jkczyz commented:
> I've pushed up a few fixes. There are some nits that could be fixed here, but I'll include them in a followup to start unblocking other PRs.
> 
> @jkczyz, if CI and you are happy, we can go ahead and get this in when you're ready. Then I'll create the follow-up issue and tag it to make sure it's a blocker for next release.

Sure, let's land this now and address the rest in follow-ups. Incremental mutants has been running for five hours, but no need to wait on it.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8112,6 +8187,309 @@ where
+									.into()))
+						},
+					}.map_err(|err| MsgHandleErrInternal::send_err_msg_no_close(format!("{}", err), msg.channel_id))?;
+					peer_state.channel_by_id.insert(channel_id, ChannelPhase::Funded(channel));
```
The fix for this actually neatens things up quite a bit in this function. Will be in the followup.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11431,39 +11791,58 @@ where
-			"Dual-funded channels not supported".to_owned(),
-			msg.channel_id.clone())), counterparty_node_id);
+		// Note that we never need to persist the updated ChannelManager for an inbound
+		// tx_complete message - interactive transaction construction does not need to
```
After consecutive `tx_complete` exchange, we will only send then initial `commitment_signed`.

We will only send a `tx_signatures` once we've received an initial `commitment_signed` if we are up first to send `tx_signatures`. Once we've received a `commitment_signed` we persist.

I've fixed this in the follow-up I'll have up soon. No PR yet.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -9618,9 +10114,12 @@ impl<'a, 'b, 'c, ES: Deref, SP: Deref> ReadableArgs<(&'a ES, &'b SP, u32, &'c Ch
 
 				blocked_monitor_updates: blocked_monitor_updates.unwrap(),
 				is_manual_broadcast: is_manual_broadcast.unwrap_or(false),
+				// If we've sent `commtiment_signed` for an interactively constructed transaction
```
Yeah this was a mistake to mark it funded at that point. Only after receiving an initial `commitment_signed` and getting a monitor do we need to persist. I'll fix the states up in the follow-up.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -9009,7 +9501,8 @@ impl<SP: Deref> Writeable for Channel<SP> where SP::Target: SignerProvider {
 			(51, is_manual_broadcast, option), // Added in 0.0.124
-			(53, funding_tx_broadcast_safe_event_emitted, option) // Added in 0.0.124
+			(53, funding_tx_broadcast_safe_event_emitted, option), // Added in 0.0.124
+			(55, self.context.next_funding_txid, option) // Added in 0.1.0
```
Removed in #3417.

dunxen replied:
The follow-up PR (not yet up) should populate it on read.
